### PR TITLE
[AMSDK-8709] - API to setting location permission level 

### PR DESCRIPTION
**Pull request by PravinPK**
Adding an API to facilitate prompting for required permission level in Android 10+ devices.

The following API is added to the Places Monitor extension.

```
public static void setLocationPermission(final PlacesMonitorLocationPermission placesMonitorLocationPermission)
```

**Sample Usage:**

To request for WhenInUse permission

```
// set the location permission
PlacesMonitor.setLocationPermission(PlacesMonitorLocationPermission.WHILE_USING_APP);
// start monitoring
PlacesMonitor.start()

```

Further when its required to upgrade to Always Allow authorization
```
// upgrade the permission level
PlacesMonitor.setLocationPermission(PlacesMonitorLocationPermission.ALLOW_ALL_TIME
);
```

## Related Issue
https://jira.corp.adobe.com/browse/AMSDK-8709

## Motivation and Context
Places monitor being flexible enough to request and upgrade permission as and when the app user needs it.

## How Has This Been Tested?

This feature is tested on emulator version (API 29, 28, 22) by calling the setLocationPermission with different values.

## Screenshots (if appropriate):

**Android 9 - Location dialog** 
Screen Shot 2019-10-02 at 11 48 53 AM

**Android 10 dialog - WhenInUse permission** 
Screen Shot 2019-10-02 at 11 53 18 AM

**Android 10 dialog - Always Allow permission** 
Screen Shot 2019-10-02 at 11 53 34 AM

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: code/places-monitor-android/src/phone/java/com/adobe/marketing/mobile/PlacesActivity.java [hunk 1]
```diff
@@ -32,18 +33,20 @@
 public class PlacesActivity extends Activity {
 
 	private static final String FINE_LOCATION = Manifest.permission.ACCESS_FINE_LOCATION;
+	private static final String BACKGROUND_LOCATION = Manifest.permission.ACCESS_BACKGROUND_LOCATION;
+	private static final String INTENT_PERMISSION_KEY = "intent permission key";
 
 	/**
-	 * Checks if the permission to access fine location is granted.
+	 * Checks if the permission to access fine location during app in use is granted.
```
Comment (sbenedicadb) [inline code/places-monitor-android/src/phone/java/com/adobe/marketing/mobile/PlacesActivity.java]: `Checks if permission to access fine location while app is in use has been granted.`

File: code/places-monitor-android/src/phone/java/com/adobe/marketing/mobile/PlacesActivity.java [hunk 2]
```diff
@@ -62,6 +65,35 @@ public static boolean isFineLocationPermissionGranted() {
 		return permissionState == PackageManager.PERMISSION_GRANTED;
 	}
 
+	/**
+	 * Checks if the permission to access fine location in background is granted.
```
Comment (sbenedicadb) [inline code/places-monitor-android/src/phone/java/com/adobe/marketing/mobile/PlacesActivity.java]: `Checks if permission to access fine location while the app is in background has been granted.`

File: code/places-monitor-android/src/phone/java/com/adobe/marketing/mobile/PlacesActivity.java [hunk 3]
```diff
@@ -70,8 +102,9 @@ public static boolean isFineLocationPermissionGranted() {
 	 *   <li> Does not invoke the permission dialog if the app context is null</li>
 	 * </ol>
 	 *
+	 * @param locationPermission The location permission setting that the user will be prompted for, could be {@link PlacesMonitorLocationPermission#WHILE_USING_APP} or {@link PlacesMonitorLocationPermission#ALLOW_ALL_TIME}
 	 */
-	public static void askPermission() {
+	public static void askPermission(PlacesMonitorLocationPermission locationPermission) {
```
Comment (sbenedicadb) [inline code/places-monitor-android/src/phone/java/com/adobe/marketing/mobile/PlacesActivity.java]: `final` parameters please

File: code/places-monitor-android/src/phone/java/com/adobe/marketing/mobile/PlacesLocationManager.java [hunk 4]
```diff
@@ -72,28 +75,42 @@
...
 	void startMonitoring() {
+		if (requestedLocationPermission == PlacesMonitorLocationPermission.WHILE_USING_APP) {
+			if (!PlacesActivity.isWhileInUsePermissionGranted()) {
+				Log.debug(PlacesMonitorConstants.LOG_TAG, "Requesting while in use location permission");
+				PlacesActivity.askPermission(requestedLocationPermission);
+				return;
+			}
+		} else if (requestedLocationPermission == PlacesMonitorLocationPermission.ALLOW_ALL_TIME) {
+			if (!PlacesActivity.isBackgroundPermissionGranted()) {
+				Log.debug(PlacesMonitorConstants.LOG_TAG, "Requesting while in use location permission");
```
Comment (sbenedicadb) [inline code/places-monitor-android/src/phone/java/com/adobe/marketing/mobile/PlacesLocationManager.java]: debug log is wrong

File: code/places-monitor-android/src/phone/java/com/adobe/marketing/mobile/PlacesLocationManager.java [hunk 5]
```diff
@@ -157,7 +174,7 @@ public void onFailure(Exception e) {
 
 					case LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE: {
 						Log.error(PlacesMonitorConstants.LOG_TAG,
-								  "Failed to start location updates, status code : SETTINGS_CHANGE_UNAVAILABLE");
+								"Failed to start location updates, status code : SETTINGS_CHANGE_UNAVAILABLE");
```
Comment (sbenedicadb) [inline code/places-monitor-android/src/phone/java/com/adobe/marketing/mobile/PlacesLocationManager.java]: i don't know what this means.

**Review by sbenedicadb — CHANGES_REQUESTED** [review]
a couple small code changes.  i'd like you to review some of the in-code docs w/ rekha for suggestions on how to make things better phrased.

File: code/places-monitor-android/src/phone/java/com/adobe/marketing/mobile/PlacesLocationManager.java (hunk 5 above)
Comment (PravinPK) [reply code/places-monitor-android/src/phone/java/com/adobe/marketing/mobile/PlacesLocationManager.java]: explained

**Review by sbenedicadb — APPROVED** [review]

**Pull request merged**